### ISSUE TITLE
Improve LSP performance by skip compiling version behind current version

### ISSUE THREAD

**Issue by logo749** [kcl-lang/kcl#1515]
## Enhancement
When file changed, the LSP compile the code in multi thread. But all thread seem to run sequentially because of the acquiring of the db write lock:
![image](https://github.com/user-attachments/assets/70fced9a-46a7-4ca2-b766-6a7ebb10d081)

I try to move the lock close to the insertion of the db, but it won't work, because it seem that we change the gs in the compile_with_params.

I propose a tiny optimization which is checking the version after we get snapshot.db write lock. In situation when user enter continuously in a big kcl file, this can skip many useless compiling of version behind current version.

I have some question:
1. Is my idea above correct?
2. Should we add debouncing when file changed?
3. Any plan to make the compile compile_with_params parallelizable by removing the db lock?

**Comment by He1pa** [comment]
debouncing is needed, and the version check is also designed for this. But I dont think you optimization will work(if you mean check version before compile). The main problem is that the compilation process now takes quite a long time. The process is as follows:
![image](https://github.com/user-attachments/assets/91df30ad-2835-4a2e-bde9-8e9ef8abce7c)

DB lock is indeed a tricky problem. We have made many changes before, but have not achieved very good results. So, if you have good suggestions and ideas, please continue to discuss.

**Comment by logo749** [comment]
Yes, when there two event, the result is the same. But when there is 3 or more event, only the first and the lastest event will be compile. And I want to add debounce to optimize the first event, so when a user press key continuely, only the lastest will handled.
![image](https://github.com/user-attachments/assets/e25e8c79-93fb-4b09-9e04-6d750d5aa806)

**Comment by logo749** [comment]
> I have also tried moving the location of db.lock, but it seems that there will be thread safety restrictions. I think it is correct to acquire the lock when need writing(after compiling), but I don't know how to implement it :)

Yes, I think this is very complicated and need deep understanding of the lsp code base...

**Comment by He1pa** [comment]
> Yes, when there two event, the result is the same. But when there is 3 or more event, only the first and the lastest event will be compile. And I want to add debounce to optimize the first event, so when a user press key continuely, only the lastest will handled. ![image](https://private-user-images.githubusercontent.com/80381139/351317689-e25e8c79-93fb-4b09-9e04-6d750d5aa806.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjE3ODg5MzksIm5iZiI6MTcyMTc4ODYzOSwicGF0aCI6Ii84MDM4MTEzOS8zNTEzMTc2ODktZTI1ZThjNzktOTNmYi00YjA5LTllMDQtNmQ3NTBkNWFhODA2LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA3MjQlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNzI0VDAyMzcxOVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTM1MjgyZTQ1ZWEzNTBlZGM4ZmRiY2RmZmYwZWVjNjRiODNjOWUzNzZhMGU3ZjJkNzhhY2ZiZDQyNThiYWMxZmUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.uvSTziquzH5bnWn8jotsK1rcmIFec9BTPjwm6ADE83A)

Ok，can you provide pseudo code? If it is very simple, it is also OK to submit a PR directly. 
Thank you very much for your contribution

**Comment by logo749** [comment]
> > Yes, when there two event, the result is the same. But when there is 3 or more event, only the first and the lastest event will be compile. And I want to add debounce to optimize the first event, so when a user press key continuely, only the lastest will handled. ![image](https://private-user-images.githubusercontent.com/80381139/351317689-e25e8c79-93fb-4b09-9e04-6d750d5aa806.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjE3ODg5MzksIm5iZiI6MTcyMTc4ODYzOSwicGF0aCI6Ii84MDM4MTEzOS8zNTEzMTc2ODktZTI1ZThjNzktOTNmYi00YjA5LTllMDQtNmQ3NTBkNWFhODA2LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA3MjQlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNzI0VDAyMzcxOVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTM1MjgyZTQ1ZWEzNTBlZGM4ZmRiY2RmZmYwZWVjNjRiODNjOWUzNzZhMGU3ZjJkNzhhY2ZiZDQyNThiYWMxZmUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.uvSTziquzH5bnWn8jotsK1rcmIFec9BTPjwm6ADE83A)
> 
> Ok，can you provide pseudo code? If it is very simple, it is also OK to submit a PR directly. Thank you very much for your contribution

Ok, I will submit a PR later :)

**Comment by He1pa** [comment]
closed by #1584
